File: Java/CircularQueue.java
package Java;

public class CircularQueue {

    static int front = -1;
    static int rear = -1;
    static int[] que = new int[5];

    static void enqueue(int x) {
        if (rear == que.length - 1 && front != 0) {
            rear = 0;
        } else if (rear == que.length - 1 || rear + 1 == front) {
            System.out.println("Overflow");
            System.exit(0);
        } else if (front == -1 && rear == -1) {
            front = rear = 0;
        } else {
            rear++;
        }
        que[rear] = x;
    }

    static void dequeue() {
        if (rear == -1 && front == -1) {
            System.out.println("Queue is already empty");
            System.exit(0);
        } else if (rear == front) {
            front = rear = -1;
            System.out.println("Deleted " + que[front - 1]);
        } else {
            front++;
            System.out.println("Deleted " + que[front - 1]);
        }
    }

    static void display() {
        if (front == -1 && rear == -1) {
            System.out.println("Queue is empty");
            System.exit(0);
        }
        if (front > rear) {
            for (int i = front; i < que.length; i++) {
                System.out.print(que[i] + " | ");
            }
            for (int j = 0; j <= rear; j++) {
                System.out.print(que[j] + " | ");
            }
            System.out.println();
        } else {
            for (int i = front; i <= rear; i++) {
                System.out.print(que[i] + " | ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        CircularQueue.enqueue(1);
        CircularQueue.enqueue(2);
        CircularQueue.enqueue(3);
        CircularQueue.enqueue(4);
        CircularQueue.enqueue(5);
        CircularQueue.dequeue();
        CircularQueue.enqueue(6);
        CircularQueue.dequeue();
        CircularQueue.enqueue(7);
        CircularQueue.display();
    }
}

File: C/SHIT DS CODE/queue/CircularQueue.c
#include <stdio.h>
#define SIZE 5

struct Queue
{
    int queue[SIZE];
    int front;
    int rear;
} *q;

void createQueue(struct Queue *q)
{
    q->front = -1;
    q->rear = -1;
}

int isEmpty(struct Queue *q)
{
    return q->front == -1;
}

int isFull(struct Queue *q)
{
    return (q->rear + 1) % SIZE == q->front;
}

void enQueue(struct Queue *q, int x)
{
    if (isFull(q))
    {
        printf("Queue is full\n");
        return;
    }
    if (isEmpty(q))
    {
        q->front = 0;
    }
    q->rear = (q->rear + 1) % SIZE;
    q->queue[q->rear] = x;
    printf("Enqueued %d in queue\n", x);
}

int deQueue(struct Queue *q)
{

    if (isEmpty(q))
    {
        printf("Queue is empty");
        return -1;
    }
    int data = q->queue[q->front];

    if (q->front == q->rear)
    {
        q->rear = -1;
        q->front = -1;
    }
    else
    {
        q->front = (q->front + 1) % SIZE;
    }

    printf("Deleted %d in queue\n", data);
    return data;
}

void display(struct Queue *q)
{
    if (isEmpty(q))
    {
        printf("Empty");
        return;
    }
    int i = q->front;
    while (i != q->rear)
    {
        printf("%d ", q->queue[i]);
        i = (i + 1) % SIZE;
    }
    printf("%d ", q->queue[q->rear]);
}

int main(int argc, char const *argv[])
{
    struct Queue q;
    createQueue(&q);
    enQueue(&q, 1);
    enQueue(&q, 2);
    enQueue(&q, 3);
    enQueue(&q, 4);
    enQueue(&q, 5);
    deQueue(&q);
    enQueue(&q, 6);
    display(&q);
    return 0;
}
